concurrency-limits-core/src/main/java/com/netflix/concurrency/limits/limit/Gradient2Limit.java

```diff
@@ -275,16 +275,24 @@ public int _update(final long startTime, final long rtt, final int inflight, fin
             this.longRtt.update(current -> current.doubleValue() * 0.95);
         }
 
-        // Don't grow the limit if we are app limited
-        if (inflight < estimatedLimit / 2) {
-            return (int) estimatedLimit;
+        // Rtt could be higher than longRtt because of smoothing longRtt updates
+        // so set to 1.0 to indicate no queuing.  Otherwise calculate the slope and don't
+        // allow it to be reduced by more than half to avoid aggressive load-shedding due to
+        // outliers. If a request is dropped, choose lowest gradient value to initiate
+        // load-shedding.
+        final double gradient = didDrop ? 0.5
+                : Math.max(0.5, Math.min(1.0, tolerance * longRtt / shortRtt));
+
+        double queueSize = 0;
+        // Don't grow the limit if not necessary
+        if (gradient == 1.0) {
+            if (inflight < estimatedLimit / 2) {
+                return (int) estimatedLimit;
+            }
+            queueSize = this.queueSize.apply((int)this.estimatedLimit);
+            queueSizeSampleListener.addSample(queueSize);
```

Comment:
Don't understand why this needs to be here. More explanation needed as part of comment in method and diff.

Reply:
gradient = 1.0 means the response times are good and system is not under load. So, we would like to increase the limit by taking queueSize into account.

Incase this caused a confusion, I need to move the comment `// Don't grow the limit if not necessary` to line 290

Reply:
Sorry, should have been more clear. Why is line 292 and 293, here? What is it solving? Not understanding why it could not be as it was previously.

Reply:
Glad you asked! queueSize is a fixed/calculated amount the estimated limit can grow while latencies remain low. Before, queueSize was taken into account even when we wanted to lower the limit. Now, we set/calculate queueSize only if latencies are low, i.e., gradient = 1.0. We can talk directly if you have any more questions.